File: laba_2/task_1/task_1.py
def longest_ones(s, K):
    left = 0
    zero_count = 0
    max_len = 0
    
    for right in range(len(s)):
        if s[right] == '0':
            zero_count += 1
        
        while zero_count > K:
            if s[left] == '0':
                zero_count -= 1
            left += 1
        
        max_len = max(max_len, right - left + 1)
    
    return max_len

def get_valid_binary_string():
    while True:
        s = input("Введите строку из 0 и 1: ").strip()
        if not s:
            print("Пустая строка! Попробуйте ещё раз.")
            continue
        if all(c in '01' for c in s):
            return s
        print("Вам нужно ввести строку из 0 и 1. Попробуйте еще раз!")

def main():
    s = get_valid_binary_string()
    
    while True:
        try:
            K = int(input("Введите K (сколько нулей можно заменить): "))
            break
        except ValueError:
            print("Ошибка! Нужно ввести ЧИСЛО. Попробуйте ещё раз: ")
    
    result = longest_ones(s, K)
    print(result)

if __name__ == "__main__":
    main()

File: laba_2/task_1/task_1.js
const readline = require('readline');

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

function ask(question) {
    return new Promise(resolve => rl.question(question, resolve));
}

function longestOnes(s, K) {
    let left = 0;
    let zeroCount = 0;
    let maxLen = 0;

    for (let right = 0; right < s.length; right++) {
        if (s[right] === '0') zeroCount++;

        while (zeroCount > K) {
            if (s[left] === '0') zeroCount--;
            left++;
        }
        maxLen = Math.max(maxLen, right - left + 1);
    }
    return maxLen;
}
async function getBinaryString() {
    while (true) {
        const input = await ask("Введите строку из 0 и 1: ");
        const s = input.trim();
      
        if (!s) {
            console.log("Пустая строка. Попробуйте ещё раз.");
            continue;
        }
        if (/^[01]+$/.test(s)) return s;
        
        console.log("Ошибка! Только символы '0' и '1'. Повтор:");
    }
}
async function getK() {
    while (true) {
        const input = await ask("Введите K (сколько нулей можно заменить): ");
        const k = Number(input.trim());
        
        if (Number.isInteger(k) && k >= 0) return k;
        
        console.log("Ошибка! Введите целое неотрицательное число.");
    }
}
async function main() {
    console.log("=== Longest Ones ===\n");
    
    const s = await getBinaryString();
    const K = await getK();
    const result = longestOnes(s, K);
    
    console.log(`\n Ответ: ${result}`);
    rl.close();
}

main();